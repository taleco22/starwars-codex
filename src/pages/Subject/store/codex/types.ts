/**
 * Action Types
 */

export enum CodexTypes {
  SET_DATA = "data/set_data",
  FAILURE = "data/failure",
  IS_LOADING = "data/is_loading",
  ASYNC_GET_DATA = "data/async_get_data",
}

/**
 * Data Types
 */

export interface Codex {
  count: number
  next: string | null
  previous: string | null
  results: any[]
}

/**
 * State Types
 */

export interface CodexState {
  data: Codex
  isLoading: boolean
}

export interface CodexAction {
  type: string
  payload: CodexState
}
