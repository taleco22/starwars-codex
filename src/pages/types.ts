import { RouteComponentProps } from "react-router-dom"

export type CategoriesProps = RouteComponentProps

export type HomeProps = RouteComponentProps

export type SParams = { theme: string }
export type SubjectProps = RouteComponentProps<SParams>

export interface SelectComponent {
  [key: string]: JSX.Element
  people: JSX.Element
  planets: JSX.Element
  vehicles: JSX.Element
  starships: JSX.Element
  species: JSX.Element
  films: JSX.Element
}

export type URL =
  | "people"
  | "planets"
  | "starships"
  | "vehicles"
  | "films"
  | "species"
