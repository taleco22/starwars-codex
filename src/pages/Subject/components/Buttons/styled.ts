import styled from "styled-components";
import {
  globalButton,
  globalItem,
  globalList,
  globalWindow,
} from "../../../../global-styled/index";

export const Window = styled(globalWindow)``;
export const List = styled(globalList)``;
export const Item = styled(globalItem)``;
export const Button = styled(globalButton)``;

export const Controls = styled.footer`
  display: flex;
  position: relative;

  @media (min-width: 768px) {
    margin-left: 10vw;
    max-width: 40%;
  }
`;

export const Page = styled.div`
  display: flex;
  position: absolute;
  top: 5rem;
  text-align: center;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-size: 1.6rem;
  text-decoration: underline;
`;
