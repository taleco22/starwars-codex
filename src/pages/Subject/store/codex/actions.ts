import { Action } from "redux";
import { action } from "typesafe-actions";
import { Codex, CodexTypes } from "./types";

export const triggerLoading = (): Action => action(CodexTypes.IS_LOADING);

export const getAsyncData = (urlPath: string) => action(CodexTypes.ASYNC_GET_DATA, { urlPath });

export const setData = (data: Codex) => action(CodexTypes.SET_DATA, { data });

export const failedRequest = (error: Error) =>
  action(CodexTypes.FAILURE, undefined, undefined, error);
