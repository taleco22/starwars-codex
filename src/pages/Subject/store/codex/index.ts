import { Reducer } from "redux";
import { CodexAction, CodexState, CodexTypes } from "./types";

export const initialState: CodexState = {
  data: {
    count: 0,
    previous: null,
    next: null,
    results: [{}, {}, {}, {}, {}],
  },
  isLoading: true,
};

export const codexReducer: Reducer<CodexState, CodexAction> = (state = initialState, action) => {
  switch (action.type) {
    case CodexTypes.SET_DATA:
      return {
        ...state,
        data: action.payload.data,
        isLoading: false,
      };
    case CodexTypes.IS_LOADING:
      return initialState;
    default:
      return initialState;
  }
};
