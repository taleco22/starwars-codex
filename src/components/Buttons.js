import React from "react";
import {useHistory} from "react-router-dom";
import * as S from "../styles/styled/index";

export default function Buttons({
  count,
  previous,
  next,
  theme,
  dispatch,
  dataCreators,
  extractPathFromUrl
}) {
  const history = useHistory();

  const goto = (url, back) => {
    if (!url) {
      if (back) history.goBack();
      return;
    }
    dispatch(dataCreators.triggerLoading());
    dispatch(dataCreators.get(extractPathFromUrl(url)));
  };

  return (
    <S.Controls>
      <S.Button off={previous === null} onClick={() => goto(previous, "back")}>
        {previous === null ? "Go Back" : "Previous"}
      </S.Button>

      <S.Button off={next === null} onClick={(e) => goto(next)}>
        Next
      </S.Button>

      <S.Page>
        {[...Array(Math.trunc(count / 10) + (count % 10 > 0 && 1))].map((v, i) => (
          <S.Button
            number
            tabIndex="0"
            key={i + 100}
            onClick={(e) => {
              goto(`https://swapi.co/api/${theme}/?page=${i + 1}`);
            }}
          >
            {i + 1}
          </S.Button>
        ))}
      </S.Page>
    </S.Controls>
  );
}
