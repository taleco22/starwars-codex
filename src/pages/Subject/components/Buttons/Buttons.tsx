import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { ApplicationState } from "../../../../redux";
import { getAsyncData, triggerLoading } from "../../../../redux/ducks/codex/actions";
import { extractPathFromUrl } from "../../../../services/api";
import { Button, Controls, Page } from "./styled";

interface ButtonProps {
  theme: string;
}

export const selector = (state: ApplicationState) => state.codex.data;

const Buttons: React.FC<ButtonProps> = ({ theme }) => {
  const { count, next, previous } = useSelector(selector);

  const history = useHistory();
  const dispatch = useDispatch();

  const goto = (url: string | null, back?: string) => {
    if (!url) {
      if (back) history.goBack();
      return;
    }
    dispatch(triggerLoading());
    dispatch(getAsyncData(extractPathFromUrl(url)));
  };

  return (
    <Controls>
      <Button off={previous === null} onClick={() => goto(previous, "back")}>
        {previous === null ? "Go Back" : "Previous"}
      </Button>

      <Button off={next === null} onClick={(e) => goto(next)}>
        Next
      </Button>

      <Page>
        {[...Array(Math.trunc(count / 10) + (count % 10 > 0 ? 1 : 0))].map((_, i) => (
          <Button
            paginate
            tabIndex={0}
            key={i + 100}
            onClick={(e) => {
              goto(`https://swapi.co/api/${theme}/?page=${i + 1}`);
            }}
          >
            {i + 1}
          </Button>
        ))}
      </Page>
    </Controls>
  );
};

export default Buttons;
