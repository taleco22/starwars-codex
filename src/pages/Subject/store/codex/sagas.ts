import { call, put, takeLatest } from "redux-saga/effects";
import { swapi } from "../../services/api";
import { failedRequest, setData } from "./actions";
import { CodexTypes } from "./types";

export function codexWatcher() {
  return takeLatest(CodexTypes.ASYNC_GET_DATA, getData);
}

interface Action {
  type: string;
  payload: { urlPath: string };
}

export function* getData(action: Action) {
  try {
    const jsonResponse = yield call(swapi, action.payload.urlPath);
    yield put(setData(jsonResponse));
  } catch (e) {
    yield put(failedRequest());
  }
}
