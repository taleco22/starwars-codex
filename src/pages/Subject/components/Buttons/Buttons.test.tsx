import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import * as React from "react";
import * as redux from "react-redux";
import configureStore from "redux-mock-store";
import { ApplicationState } from "../../../../redux";
import { CodexState } from "../../../../redux/ducks/codex/types";
import Buttons, { selector } from "./Buttons";

const mockGoBack = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    goBack: mockGoBack,
  }),
}));

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

const use_selector = jest.spyOn(redux, "useSelector");
const use_dispatcher = jest.spyOn(redux, "useDispatch");

const initial_state = {
  data: { count: 10, previous: "previous", next: "next", results: [{}] },
  isLoading: false,
};

const nullState = {
  data: { count: 1, previous: null, next: null, results: [{}] },
  isLoading: false,
};

const reducer = (state: CodexState) => ({ codex: state });
const mockStore = configureStore<ApplicationState>()(reducer(initial_state));
const mockEmpty = configureStore<ApplicationState>()(reducer(nullState));

describe("Buttons.tsx", () => {
  describe("store on", () => {
    let result: { container: HTMLElement };

    beforeEach(async () => {
      result = render(
        <redux.Provider store={mockStore}>
          <Buttons theme="people" />
        </redux.Provider>
      );
    });

    afterEach(() => {
      use_dispatcher.mockClear();
      mockDispatch.mockClear();
    });

    it("render component correctly", () => {
      expect(result.container).toMatchSnapshot();
      expect(use_dispatcher).toBeCalled();
    });

    it("get data from selector", () => {
      expect(use_dispatcher).toBeCalled();
      expect(selector(mockStore.getState()));
    });

    it("trigger onClick handler", () => {
      const button = screen.getByText(/Next/g);
      fireEvent.click(button);
      expect(use_dispatcher).toBeCalledTimes(1);
      expect(mockGoBack).not.toBeCalled();
      expect(mockDispatch).toBeCalledTimes(2);
    });

    it("trigger onClick handler", () => {
      const button = screen.getByText(/\d/g);
      fireEvent.click(button);
      expect(use_dispatcher).toBeCalledTimes(1);
      expect(mockDispatch).toBeCalledTimes(2);
    });
  });

  describe("store off", () => {
    let result: { container: HTMLElement };

    beforeEach(async () => {
      result = render(
        <redux.Provider store={mockEmpty}>
          <Buttons theme="people" />
        </redux.Provider>
      );
    });

    afterEach(() => {
      use_dispatcher.mockClear();
      mockGoBack.mockClear();
    });

    it("renders", () => {
      expect(result.container).toMatchSnapshot();
    });

    it("trigger onClick handler", () => {
      const button = screen.getByText(/Go Back/g);
      fireEvent.click(button);
      expect(mockGoBack).toBeCalled();
      expect(mockDispatch).not.toBeCalled();
    });

    it("trigger onClick handler", () => {
      const button = screen.getByText(/Next/g);
      fireEvent.click(button);
      expect(mockGoBack).not.toBeCalled();
      expect(mockDispatch).not.toBeCalled();
    });
  });
});
