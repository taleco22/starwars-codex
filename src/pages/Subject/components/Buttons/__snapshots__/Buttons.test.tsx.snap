// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons.tsx store off renders 1`] = `
<div>
  <footer
    class="sc-fzozJi bBjIJb"
  >
    <button
      class="sc-AxheI fOcLmV"
    >
      Go Back
    </button>
    <button
      class="sc-AxheI fOcLmV"
    >
      Next
    </button>
    <div
      class="sc-fzoLsD wRDqf"
    >
      <button
        class="sc-AxheI fEbfFm"
        tabindex="0"
      >
        1
      </button>
    </div>
  </footer>
</div>
`;

exports[`Buttons.tsx store on render component correctly 1`] = `
<div>
  <footer
    class="sc-fzozJi bBjIJb"
  >
    <button
      class="sc-AxheI jfMBhg"
    >
      Previous
    </button>
    <button
      class="sc-AxheI jfMBhg"
    >
      Next
    </button>
    <div
      class="sc-fzoLsD wRDqf"
    >
      <button
        class="sc-AxheI fEbfFm"
        tabindex="0"
      >
        1
      </button>
    </div>
  </footer>
</div>
`;
